src/twisted/logger/_logger.py

```diff
@@ -92,16 +93,13 @@ def hello(self):
         C{Something}, and C{Something().log.source} would be an instance of
         C{Something}.
         """
-        if oself is None:
-            source = type
-        else:
-            source = oself
+        if not self._isDescriptor:
+            self.namespace = ".".join([type.__module__, type.__name__])
+            self._isDescriptor = True
 
-        return self.__class__(
-            ".".join([type.__module__, type.__name__]),
-            source,
-            observer=self.observer,
-        )
+        self.source = type if oself is None else oself
```

Comment:
The descriptor is declared at the class scope.

Which means that this change is potentially caching an object that just tried to log something *forever*, regardless of the size of that object; which could make for an extremely pernicious memory leak.

We need a different implementation strategy here, which doesn't cache any instance-level stuff at module-level by accident!

Reply:
@glyph I'm afraid you lost me here, can you please clarify a bit?  What is being cached at module level?  As far as I understand (and my understanding may be incorrect, of course), all I am doing here is telling an instance of Logger to not create a new Logger instance every time someone calls .info/.debug/.error/.warning/.failure on this instance.

I'm happy to fix the problem you've described, but I don't actually see the problem, so any light you can shed on it will help :)

Reply:
I should probably also add that I've been running this code in production for a month now, and it does what I expect, at least for my application: memory usage is down, and performance is slightly improved due to reduced object-creation overhead.

Reply:
Here's an example of what I mean.

```python3
from gc import collect
from twisted.logger import Logger, globalLogBeginner, textFileLogObserver
import sys

class MyLogging(object):
    log = Logger()

    def __del__(self):
        print("bye!")

globalLogBeginner.beginLoggingTo([textFileLogObserver(sys.stdout)])
print("start")
MyLogging().log.info("logged!")
collect()
print("end")
```

On current `trunk`, this logs:

```
2019-11-02T16:05:28-0700 [stdout#info] start
2019-11-02T16:05:28-0700 [__main__.MyLogging#info] logged!
2019-11-02T16:05:28-0700 [stdout#info] bye!
2019-11-02T16:05:28-0700 [stdout#info] end
```

and with this change, it does this instead:

```
2019-11-02T16:06:36-0700 [stdout#info] start
2019-11-02T16:06:36-0700 [__main__.MyLogging#info] logged!
2019-11-02T16:06:36-0700 [stdout#info] end
bye!
```

The reason being that `Logger` is instantiated at class scope, and maintains a reference to the last instance that touched it, forever.  This was previously not the case.  So there's a lifecycle bug here.

This also introduces a thread-safety concern, because multiple loggers on different instances might end up getting the wrong result for `source`.

Reply:
Ah, ok, I get it now, thank you very much for taking the time to explain it.  As an aside, this is why I prefer module-level logging, and its much lower level of shenanigans ;)

I'll see if I can figure out something better.  Relatedly, since Pypy is GC-only, we may get different behavior there as well.

Reply:
So, at least on Python 3.6.9, the call to collect() doesn't matter (maybe it does on Pypy); omitting it changes nothing about the example's output.

Changing the descriptor to "self.source = weakref.ref(source)" ensures that the del method is called (does anyone actually use del anymore?), but the log output changes slightly:

```python
python ./9680.py 
2019-11-04T14:57:43-0600 [stdout#info] start
2019-11-04T14:57:43-0600 [stdout#info] bye!
2019-11-04T14:57:43-0600 [__main__.MyLogging#info] logged!
2019-11-04T14:57:43-0600 [stdout#info] end
```

I don't think this ordering matters much in this case.

Another option is always setting source to be the class, not the instance, since that is what gets emitted in the log anyway.  I grepped the codebase and the only thing in Twisted that cares if the source is an instance or a class is some test suite code, but I don't know how many people out in the world rely on the source attribute to be an instance when performing logging.  I suspect in general that people just log their message and that's that, in which case changing self.source to always be the class would "just work".

Regarding thread-safety: given 3 instances of a class, running in three threads, each instance calling self.log.info('blah'), maybe that use-case means we should create exactly one Logger instance per instance of the class, so that each instance has its own Logger instance, but can use that one Logger instance rather than creating a new one upon each invocation.

Reply:
The whole point of binding the instance is *to bind the instance* so you can do stuff like `self.log.info("hey buddy {something that uses attributes of log_source}")` :). If you want to set it to the class, do `MyLogging.log.info()`; that also works. Or just use module-level loggers!

It might be interesting to reduce object creation or do some lazy caching in the case of intensive use of this feature, but it sounds like the idiom for the most part should be "don't use that feature much, unless you really want what it gives you".

So: module-level loggers for everybody :).